### PR TITLE
Move extended tests to separate Go packages

### PR DESCRIPTION
**Pull request by mfojtik**
This patch will move the test files into their Go packages:

* `./test/extended/images` is a package that holds all test cases related to testing different Docker images (including mysql, sti images, deployer, etc...)
* `./test/extended/builds` is a package that holds all test cases related to builds

Note that the test files does not have '_test.go' suffix. This is copying the upstream pattern and allow to "import" that files in "extended_test.go" which is a runner for all extended tests. (The `_test.go` files are not 'importable' as packages in Go)

Also note that the Go package names are not corresponding to the extended test groups, which is intentional.
You can have `image tests` that you want to execute against the "default" server configuration, but you can also have image test that requires different server configuration (iow. different 'group'). (Explained here: https://github.com/mfojtik/origin/blob/extended-docs/docs/writing-extended-test.md#groups-vs-packages)

To divide test cases into groups, you have to use the `default:` prefix in the `Describe()` statements for particular test case you want to execute in that group.

### CODE REVIEW

File: hack/test-extended/default/run.sh [hunk 1]
```diff
@@ -144,12 +145,13 @@ oc create -n openshift -f examples/image-streams/image-streams-centos7.json --co
...
 echo "[INFO] MASTER IP - ${MASTER_ADDR}"
 echo "[INFO] SERVER CONFIG PATH - ${SERVER_CONFIG_DIR}"
+echo "[INFO] Starting extended tests ..."
 
 # Run the tests
-KUBECONFIG="${ADMIN_KUBECONFIG}" \
-  ginkgo -progress -focus="default:" -p ./test/extended/extended.test
+pushd ${OS_ROOT}/test/extended >/dev/null
+export KUBECONFIG="${ADMIN_KUBECONFIG}"
+export EXTENDED_TEST_PATH="${OS_ROOT}/test/extended"
```
Comment (mfojtik) [inline hack/test-extended/default/run.sh]: @deads2k before you ask, this is another caveat of Ginkgo parallel test execution... it changes PWD for 'nodes' that run the test.. so the first node PWD is correct, but the second is reset to `_output/bin/`....

I fixed this by explicitly telling where is the root folder for extended tests, so the test cases can rely on it when reading the fixture files.

**Comment by mfojtik** [comment]
@deads2k @jhadvig @mnagy PTAL

[test][extended:all]

File: hack/test-extended/default/run.sh [hunk 2]
```diff
@@ -39,8 +39,9 @@ set +e
 
 # Compile the extended tests first to avoid waiting for OpenShift server to
 # start and fail sooner on compilation errors.
+echo "[INFO] Compiling test/extended package ..."
 GOPATH="${OS_ROOT}/Godeps/_workspace:${GOPATH}" \
-  ginkgo build -r ./test/extended 
+  go test -c ./test/extended -o ${OS_OUTPUT_BINPATH}/extended.test || exit 1
```
Comment (mfojtik) [inline hack/test-extended/default/run.sh]: @deads2k I moved this up so it fails sooner if you make a typo when writing extended tests... waiting ~2min for openshift to start to realize that you forgot to close string is frustrating ;D

File: test/extended/extended_test.go [hunk 3]
```diff
@@ -37,11 +42,11 @@ func init() {
 	extendedOutputDir := filepath.Join(os.TempDir(), "openshift-extended-tests")
 	os.MkdirAll(extendedOutputDir, 0600)
 
-	flag.StringVar(&testContext.KubeConfig, clientcmd.RecommendedConfigPathFlag, kubeConfigPath(), "Path to kubeconfig containing embeded authinfo.")
-	flag.StringVar(&testContext.OutputDir, "extended-tests-output-dir", extendedOutputDir, "Output directory for interesting/useful test data, like performance data, benchmarks, and other metrics.")
+	flag.StringVar(&exutil.TestContext.KubeConfig, clientcmd.RecommendedConfigPathFlag, exutil.KubeConfigPath(), "Path to kubeconfig containing embeded authinfo.")
```
Comment (jhadvig) [inline test/extended/extended_test.go]: s/embeded/embedded

File: test/extended/images/mysql_ephemeral.go [hunk 4]
```diff
@@ -13,12 +12,12 @@ import (
 var _ = Describe("default: MySQL ephemeral template", func() {
 	defer GinkgoRecover()
 	var (
-		templatePath = filepath.Join("..", "..", "examples", "db-templates", "mysql-ephemeral-template.json")
-		oc           = exutil.NewCLI("mysql-create", kubeConfigPath())
+		templatePath = exutil.FixturePath("..", "..", "examples", "db-templates", "mysql-ephemeral-template.json")
```
Comment (jhadvig) [inline test/extended/images/mysql_ephemeral.go]: :+1:

**Comment by jhadvig** [comment]
@mfojtik just a small nit, otherwise LGTM
Good job :+1:

**Comment by deads2k** [comment]
lgtm.  Will a merge tag with [extended:all] require the extended tests to pass before merging?

**Comment by liggitt** [comment]
No

**Comment by liggitt** [comment]
Though that would be super useful. @danmcp ^^

**Comment by deads2k** [comment]
> Though that would be super useful. @danmcp ^^

I think wiring it up that way would preserve tagger intent.  If someone requested extended tests, they clearly thought the change was risky enough to warrant the time spent running the extended tests.

Out curiosity, have we also considered running multiple parallel merge jobs for a single PR based on different branches based on optimistic merge ordering in the queue?  Even just going one job deep would help with latency.

**Comment by mfojtik** [comment]
@deads2k I thought that if you tag already to run extended and then you do merge, it will run extended prior to merge

**Comment by deads2k** [comment]
test flake: https://github.com/openshift/origin/issues/4262 

re[test]

File: hack/test-extended/default/run.sh (hunk 2 above)
Comment (stevekuznetsov) [inline hack/test-extended/default/run.sh]: :+1:

**Comment by deads2k** [comment]
lgtm [merge]

**Comment by stevekuznetsov** [comment]
Just to be clear, right now `run.sh` has the `-focus="default:"` hard-coded in, and the code in `hack/test-extended.sh` chooses  which `run.sh` to run based on the argument given.

Currently we just have the default package, and I don't see why we need to default to this scheme. I'm adding a test right now and in order to run it and only it, I need to set up my own package, my own (copy-pasted) `run.sh` and what for? To change the string in `-focus` ? We should pass *that* in from the invocation in `hack/test-extended.sh`, or at least allow `hack/test-extended.sh default <something that will be prepended>` or passed into `grep -E` like in `hack/test-integration.sh`.

**Comment by mfojtik** [comment]
@stevekuznetsov agree, I think we should have FOCUS variable you can set to override the default:.
That will allow to just call the original default/run.sh with FOCUS set to whatever test group.

**Comment by mfojtik** [comment]
@stevekuznetsov can you make a PR for that?

**Comment by stevekuznetsov** [comment]
Sure, issue at #4266

**Comment by deads2k** [comment]
> Currently we just have the default package, and I don't see why we need to default to this scheme. I'm adding a test right now and in order to run it and only it, I need to set up my own package, my own (copy-pasted) run.sh and what for? To change the string in -focus ? We should pass that in from the invocation in hack/test-extended.sh, or at least allow hack/test-extended.sh default <something that will be prepended> or passed into grep -E like in hack/test-integration.sh.

Just to be sure you're complaining about the correct thing.  The reason you have to create your own folder and your own `run.sh` is so that we selectively choose which bucket of extended tests to run.  Buckets are likely to follow functional areas like builds, images, deployments, authentication, and authorization.  Default is supposed to a smattering of the above.  Since the initialization and process of running the tests looks different in different functional areas, each bucket is responsible to choosing how to run its tests.  Hence their own `run.sh`

**Comment by stevekuznetsov** [comment]
@deads2k A better set-up (and one I think @mfojtik was working towards in this pull) is to have a package-specific setup script (inside of `test/extended/builds`, for example) and then let the user choose which *tests* to run. We can also namespace tests (`default: build: <name>`) so that your filter string could catch functional areas as well. But there's no reason to not be able to run a specific extended test without all the rest in it's functional group.

**Pull request merged**